src/atc/cache/CachedLoader.py

```diff
@@ -0,0 +1,215 @@
+import sys
+from random import randint
+from typing import List, Optional
+
+import pyspark.sql.functions as f
+from pyspark.sql import DataFrame
+
+from atc.etl import Loader
+from atc.spark import Spark
+
+from .CachedLoaderParameters import CachedLoaderParameters
+
+
+class CachedLoader(Loader):
+    """
+    Remember to override self.write_operation and/or self.delete_operation
+    Any variable rows need to be added in this function.
+    """
+
+    class ReductionResult:
+        to_be_written: DataFrame
+        to_be_deleted: DataFrame
+
+    def __init__(self, params: CachedLoaderParameters):
+        super().__init__()
+        self.params = params
+        self.validate()
+
+    def validate(self):
+        # validate cache table schema
+        p = self.params
+        df = Spark.get().table(p.cache_table_name)
+        try:
+
+            assert set(df.columns) == set(
+                p.key_cols + p.cache_id_cols + [p.rowHash, p.loadedTime, p.deletedTime]
+            )
+        except AssertionError:
+            print(
+                "ERROR: The cache table needs to have precisely the correct schema."
+                " Found:",
+                df.columns,
+                file=sys.stderr,
+            )
+            raise
+
+        # validate overloading
+        write_overloaded = "write_operation" in self.__class__.__dict__
+        delete_overloaded = "delete_operation" in self.__class__.__dict__
+        if not (write_overloaded or delete_overloaded):
+            raise AssertionError(
+                "overload of write_operation " "or delete_operation required"
+            )
+
+        if self.__class__ is CachedLoader:
+            raise AssertionError("You should inherit from this class")
+
+    def write_operation(self, df: DataFrame) -> Optional[DataFrame]:
+        """Abstract Method to be overridden in child."""
+        return None
+
+    def delete_operation(self, df: DataFrame) -> Optional[DataFrame]:
+        """Abstract Method to be overridden in child."""
+        return None
+
+    def save(self, df: DataFrame) -> None:
+        in_cols = df.columns
+
+        if not set(self.params.key_cols).issubset(in_cols):
+            raise AssertionError(
+                "The key columns must be given in the input dataframe."
+            )
+
+        # ensure that no duplicates exist. The keys have to be a unique set.
+        df = df.dropDuplicates(self.params.key_cols)
```

Comment:
dropDuplicates is not always consistent in the way, that the duplicate chosen can vary.

Reply:
Indeed, and you should always make sure your primary keys are unique before you use the CachedLoader. This is just an insurance against corrupting the tables.